src/ray/gcs/gcs_server/gcs_task_manager.cc

```diff
@@ -137,6 +270,27 @@ GcsTaskManager::GcsTaskManagerStorage::AddOrReplaceTaskEvent(
     }
     task_to_task_attempt_index_[task_id].insert(task_attempt);
 
+    // Update the parent <-> children mapping for the removed one.
+    // Remove it's relationship with the parent.
+    auto replaced_parent_task_id = GetParentTaskId(replaced);
+    if (!replaced_parent_task_id.IsNil()) {
+      // Remove itself from it's parent's children set if any.
+      auto sibling_range =
+          parent_to_children_task_index_.equal_range(replaced_parent_task_id.Binary());
+
+      // Look for the replaced task among all the children of the parent, and erase it.
+      for (auto child_itr = sibling_range.first; child_itr != sibling_range.second;
```

Comment:
isn't this N^2? For example, if you have 20K tasks submitted from drivers, this can require 10K^2 iteration in the worst case?

Reply:
Ah good catch - yeah, I guess using a map<taskid, set<taskid>> is probably necessary then.

Reply:
resolved offline